react/features/base/participants/reducer.js

```diff
@@ -64,62 +63,81 @@ function _participant(state, action) {
+
+            if (!newState.avatar) {
+                newState.avatar = getAvatarURL(newState);
+            }
+
+            return newState;
         }
         break;
 
     case PARTICIPANT_JOINED: {
         const participant = action.participant; // eslint-disable-line no-shadow
+        const { avatar, dominantSpeaker, email, local, pinned, role }
+            = participant;
+        let { id, name } = participant;
+
+        // id
+        //
         // XXX The situation of not having an ID for a remote participant should
         // not happen. Maybe we should raise an error in this case or generate a
         // random ID.
-        const id
-            = participant.id
-                || (participant.local && LOCAL_PARTICIPANT_DEFAULT_ID);
-        const avatar
-            = participant.avatar || _getAvatarURL(id, participant.email);
+        if (!id && local) {
+            id = LOCAL_PARTICIPANT_DEFAULT_ID;
+        }
 
-        // TODO Get these names from config/localized.
-        const name
-            = participant.name || (participant.local ? 'me' : 'Fellow Jitster');
+        // name
+        if (!name) {
+            // TODO Get the from config and/or localized.
+            name = local ? 'me' : 'Fellow Jitster';
+        }
 
-        return {
+        const newState = {
             avatar,
-            email: participant.email,
+            dominantSpeaker: dominantSpeaker || false,
+            email,
             id,
-            local: participant.local || false,
+            local: local || false,
             name,
-            pinned: participant.pinned || false,
-            role: participant.role || PARTICIPANT_ROLE.NONE,
-            dominantSpeaker: participant.dominantSpeaker || false
+            pinned: pinned || false,
+            role: role || PARTICIPANT_ROLE.NONE
         };
+
+        if (!newState.avatar) {
+            newState.avatar = getAvatarURL(newState);
+        }
+
+        return newState;
     }
 
-    case PARTICIPANT_UPDATED:
-        if (state.id === action.participant.id) {
+    case PARTICIPANT_UPDATED: {
+        const participant = action.participant; // eslint-disable-line no-shadow
+        const { id } = participant;
+
+        if (state.id === id) {
             const newState = { ...state };
 
-            for (const key in action.participant) {
-                if (action.participant.hasOwnProperty(key)
+            for (const key in participant) {
+                if (participant.hasOwnProperty(key)
                         && PARTICIPANT_PROPS_TO_OMIT_WHEN_UPDATE.indexOf(key)
                             === -1) {
-                    newState[key] = action.participant[key];
+                    newState[key] = participant[key];
                 }
             }
 
             if (!newState.avatar) {
-                newState.avatar
-                    = _getAvatarURL(action.participant.id, newState.email);
+                newState.avatar = getAvatarURL(newState);
```

Comment:
I changed your source code but I don't think I changed the behavior here. What I'm thinking now though is that `getAvatarURL` may take into account an update property such as `avatarID`, `avatarURL`, or `email` which means that the new `avatar` may be different from the old `avatar` because of the updated property.